packages/plugins/inline-trace/src/index.ts

```diff
@@ -0,0 +1,313 @@
+        stopped: false,
+      })
+    },
+    onResolverCalled({ context: { request }, info }) {
+      const ctx = ctxForReq.get(request)
+      if (!ctx) return
+
+      // result was already shipped (see InlineTraceContext.stopped)
+      if (ctx.stopped) {
+        return () => {
+          // noop
+        }
+      }
+
+      const node = newTraceNode(ctx, info.path)
+      node.type = info.returnType.toString()
+      node.parentType = info.parentType.toString()
+      node.startTime = hrTimeToDurationInNanos(process.hrtime(ctx.startHrTime))
+      if (
+        typeof info.path.key === 'string' &&
+        info.path.key !== info.fieldName
+      ) {
+        // field was aliased, send the original field name too
+        node.originalFieldName = info.fieldName
+      }
+
+      return () => {
+        node.endTime = hrTimeToDurationInNanos(process.hrtime(ctx.startHrTime))
+      }
+    },
+    onParse() {
+      return ({ context: { request }, result }) => {
+        const ctx = ctxForReq.get(request)
+        if (!ctx) return
+
+        if (result instanceof GraphQLError) {
+          handleErrors(ctx, [result], options.rewriteError)
+        } else if (result instanceof Error) {
+          handleErrors(
+            ctx,
+            [
+              new GraphQLError(result.message, {
+                originalError: result,
+              }),
+            ],
+            options.rewriteError,
+          )
+        }
+      }
+    },
+    onValidate() {
+      return ({ context: { request }, result: errors }) => {
+        if (errors.length) {
+          const ctx = ctxForReq.get(request)
+          if (ctx) handleErrors(ctx, errors, options.rewriteError)
+        }
+      }
+    },
+    onExecute() {
+      return {
+        onExecuteDone({
+          args: {
+            contextValue: { request },
+          },
+          result,
+        }) {
+          // TODO: should handle streaming results? how?
+          if (!isAsyncIterable(result) && result.errors?.length) {
+            const ctx = ctxForReq.get(request)
+            if (ctx) handleErrors(ctx, result.errors, options.rewriteError)
+          }
+        },
+      }
+    },
+    onResultProcess({ request, result }) {
+      const ctx = ctxForReq.get(request)
+      if (!ctx) return
+
+      // TODO: should handle streaming results? how?
+      if (isAsyncIterable(result)) return
+
+      if (result.extensions?.ftv1 !== undefined) {
+        throw new Error('The `ftv1` extension is already present')
+      }
+
+      // onResultProcess will be called only once since we disallow async iterables
+      if (ctx.stopped) throw new Error('Trace stopped multiple times')
+
+      ctx.stopped = true
+      ctx.trace.durationNs = hrTimeToDurationInNanos(
+        process.hrtime(ctx.startHrTime),
+      )
+      ctx.trace.endTime = nowTimestamp()
+
+      const encodedUint8Array = Trace.encode(ctx.trace).finish()
+      const encodedBuffer = Buffer.from(
```

Comment:
Do you think we can use `TextDecoder` to make this platform agnostic?

Reply:
I just thought `TextDecoder` as an alternative for `Buffer.from` and `toString` combination;
```ts
const textDecoder = new TextDecoder();
const str = textDecoder.decode(encodedUint8Array);
```

Reply:
Ah, my bad. I'll see what I can do.

Reply:
The buffer is later encoded to a `base64` string, it's much easier (and more readable) to do it with a Buffer vs. [with a TextEncoder](https://stackoverflow.com/a/12713326/3633671).

Reply:
Great point by @ardatan, Buffer is exclusive to Node. Fixed with: https://github.com/dotansimha/graphql-yoga/pull/1621/commits/901ce49c786f2899111c6b9eec50e3924b88f602.